jetty-client/src/main/java/org/eclipse/jetty/client/HttpReceiver.java

```diff
@@ -591,46 +647,151 @@ public String toString()
         FAILURE
     }
 
-    private class Decoder extends IteratingNestedCallback
+    private class ContentListeners
+    {
+        private final Map<Object, Long> demands = new ConcurrentHashMap<>();
+        private final LongConsumer demand = HttpReceiver.this::demand;
+        private final List<Response.DemandedContentListener> listeners;
+
+        private ContentListeners(List<Response.ResponseListener> responseListeners)
+        {
+            listeners = responseListeners.stream()
+                .filter(Response.DemandedContentListener.class::isInstance)
+                .map(Response.DemandedContentListener.class::cast)
+                .collect(Collectors.toList());
+        }
+
+        private boolean isEmpty()
+        {
+            return listeners.isEmpty();
+        }
+
+        private void notifyBeforeContent(HttpResponse response)
+        {
+            if (isEmpty())
+            {
+                // If no listeners, we want to proceed and consume any content.
+                demand.accept(1);
+            }
+            else
+            {
+                ResponseNotifier notifier = getHttpDestination().getResponseNotifier();
+                notifier.notifyBeforeContent(response, this::demand, listeners);
+            }
+        }
+
+        private void notifyContent(HttpResponse response, ByteBuffer buffer, Callback callback)
+        {
+            if (hasManyListeners())
+                demands.replaceAll((k, v) -> v - 1);
```

Comment:
rather than decrementing the mapped demands 1 by 1 here, I think minDemand should be removed in accept below.